helusers/user_utils.py

```diff
@@ -107,7 +106,7 @@ def get_or_create_user(payload, oidc=False):
     user_id = payload.get('sub')
     if not user_id:
         msg = _('Invalid payload. sub missing')
-        raise exceptions.AuthenticationFailed(msg)
+        raise ValueError(msg)
```

Comment:
According to my analysis, this exception will end up in at least these contexts:
- when running the social auth pipeline
- as a USER_RESOLVER (called eg. from the old DRF auth classes)
- used as part of a DRF authentication classes

I think (but am not sure) that the ValueError will propagate to the top level and cause a 500, instead of the proper 40x error we had previously. It's also important to log all failed auth attempts, which is another reason to use specific exceptions (or log the error here)

Reply:
The `ValueError` is turned into DRF's `AuthenticationFailed` in [`oidc.ApiTokenAuthentication`](https://github.com/City-of-Helsinki/django-helusers/pull/33/commits/0ecf259ac5540558428d70a266d880d96533f704#diff-33dfa22394312a69b490ba7760a7c35ee221c0ac890fbd364d9783dd89de31ceR52-R55) and in [`jwt.JWTAuthentication`](https://github.com/City-of-Helsinki/django-helusers/pull/33/commits/0ecf259ac5540558428d70a266d880d96533f704#diff-3063b6feea6b8068dd059378dffe9e355fb759fdbc57780a4fe7f1e3900c9710R34-R37).

But the social auth pipeline case is not yet handled. I'm not sure if raising DRF exceptions from the social auth pipeline has ever been the correct thing to do anyway, because the social auth pipeline shouldn't probably depend on DRF either. Any idea what the pipeline should do? The documentation doesn't say much about it. It [mentions](https://python-social-auth.readthedocs.io/en/latest/developer_intro.html#understanding-the-pipeline) that:

> 2. If the function returns a dictionary or something False-ish [--] call the next step in the pipeline [--]
> 3. If something else is returned (for example, a subclass of HttpResponse), then return that to the browser.

In order to interrupt the pipeline, step 3 should be taken. But what that _something else_ could be in this case? `Django-helusers` has no idea in what kind of project it is used in and what kind of responses should be returned to the browser. The documentation doesn't say anything about how exceptions raised in a pipeline function are handled (if at all). Supposedly pipeline functions shouldn't raise any exceptions?

I suppose this hasn't been thought about before. In the default pipeline that `django-helusers` provides, `ensure_uid_is_uuid` ensures that `uid` is set (and is a UUID). Then `create_or_update_user` puts `uid` into the `sub` field. Then `get_or_create_user` is called and because `sub` field is now always set, it doesn't raise that `ValueError` (or previously the DRF `AuthenticationFailed`). So...there's no problem 😲 ?

Reply:
Ok, good enough for me! I suppose pipeline functions should catch all (known) exceptions and return an appropriate HTTP error as a HttpResponse. But in this case it seems even that is not needed.